src/contracts/core/Slasher.sol

```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
             uint64 scalingFactorAfter = SlashingAccountingUtils.findNewScalingFactor({
                 scalingFactorBefore: scalingFactorBefore,
-                rateToSlash: slashingRequest.slashingRate
+                rateToSlash: slashingRate
             });
             // update storage to reflect the slashing
-            slashedEpochHistory[operator][strategy].push(epoch);
+            slashingEpochHistory[operator][strategy].push(epoch);
+            slashingUpdates[operator][strategy][epoch] = SlashingUpdate({
+                slashingRate: slashingRate,
+                scalingFactor: scalingFactorAfter
+            });
             _shareScalingFactor[operator][strategy] = scalingFactorAfter;
             // TODO DONE?: note that this behavior risks off-by-one errors. it needs to be clearly defined precisely how the historical storage is supposed to work
-            slashingRequest.scalingFactor = scalingFactorAfter;
 
-            emit SlashingExecuted(epoch, operator, strategy, slashingRequest.slashingRate);
         }
     }
 
-    /// VIEW
+    /*******************************************************************************
+                                 INTERNAL FUNCTIONS
+    *******************************************************************************/
 
-    /**
-     * @notice fetches the requested parts per hundred million to slash for the
-     * given operator, strategy, epoch, and operator set
-     *
-     * @param operator the operator to get the requested slashing rate for
-     * @param strategy the strategy to get the requested slashing rate for
-     * @param operatorSet the operator set to get the requested requested slashing rate for
-     * @param epoch the epoch to get the requested slashing rate  for
-     *
-     * @return the requested parts per hundred million to slash for the given
-     * operator, strategy, epoch, and operator set
-     *
-     * @dev may exceed the AVS operator set's allowed slashing per epoch;
-     * the `getPendingSlashedPPHM` will accurately reflect this ceiling though.
-     */
-    function getRequestedSlashingRate(
+    /// @notice push or update the latest magnitude update for the given operator, strategy, and operatorSet
+    /// depending on the current effectTimestamp
+    function _updateMagnitudeHistory(
         address operator,
         IStrategy strategy,
-        IOperatorSetManager.OperatorSet calldata operatorSet,
-        uint32 epoch
-    ) public view returns (uint32) {
-        return requestedSlashedBips[operator][strategy][epoch][_hashOperatorSet(operatorSet)]
-            * operatorSetManager.getSlashableBips(operator, operatorSet, strategy, epoch);
+        OperatorSet calldata operatorSet,
+        MagnitudeUpdate storage magnitudeUpdate,
+        uint64 newMagnitude,
+        uint32 effectTimestamp
+    ) internal {
+        if (magnitudeUpdate.timestamp == effectTimestamp) {
```

Comment:
i feel like making this check earlier and reverting if called twice in a block is better? idk